src/consensus/pool/finality_tracker.rs

```diff
@@ -0,0 +1,305 @@
+                | FinalizationStatus::ImplicitlyFinalized(_)
+                | FinalizationStatus::ImplicitlySkipped => event,
+                FinalizationStatus::Finalized => {
+                    self.status
+                        .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+                    self.handle_finalized((slot, block_hash), &mut event);
+                    event
+                }
+            }
+        } else {
+            self.status
+                .insert(slot, FinalizationStatus::Notarized(block_hash));
+            event
+        }
+    }
+
+    /// Marks the given slot as finalized.
+    ///
+    /// Handles possibly resulting direct finalization of a block in this slot.
+    /// Further, also handles any possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_finalized(&mut self, slot: Slot) -> FinalizationEvent {
+        let mut event = FinalizationEvent::default();
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::Finalized
+                | FinalizationStatus::FinalizedAndNotarized(_)
+                | FinalizationStatus::ImplicitlyFinalized(_) => event,
+                FinalizationStatus::Notarized(block_hash) => {
+                    let block_hash = *block_hash;
+                    self.status
+                        .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+                    self.handle_finalized((slot, block_hash), &mut event);
+                    self.highest_finalized = slot.max(self.highest_finalized);
+                    event
+                }
+                FinalizationStatus::ImplicitlySkipped => unreachable!("consensus safety violation"),
+            }
+        } else {
+            self.status.insert(slot, FinalizationStatus::Finalized);
+            self.highest_finalized = slot.max(self.highest_finalized);
+            event
+        }
+    }
+
+    /// Returns the highest finalized slot.
+    ///
+    /// Does not consider whether we have the block, or even know the hash.
+    pub fn highest_finalized(&self) -> Slot {
+        self.highest_finalized
+    }
+
+    /// Handles the direct finalization of the given block.
+    ///
+    /// Recurses through ancestors, potentially implicitly finalizing them.
+    ///
+    /// Updates the `event` all along the way with:
+    /// - The finalized block,
+    /// - any potentially implicitly finalized blocks, AND
+    /// - any implicitly skipped slots.
+    fn handle_finalized(&mut self, finalized: BlockId, event: &mut FinalizationEvent) {
+        event.finalized = Some(finalized);
+        if let Some(parent) = self.parents.get(&finalized) {
+            self.handle_implicitly_finalized(finalized.0, *parent, event);
+        }
+    }
+
+    /// Handles the indirect finalization of the given block.
```

Comment:
If the node joined the network after being absent for a long time, this function could recurse a lot.  There might be a potential for running out of stack space.  Probably not something we need to worry about in this PR but could be something to track in an issue at least.